### PR TITLE
[13.0][MIG] account_invoice_comment_template: Migration to 13.0

### PR DESCRIPTION
**Pull request by victoralmau**
Migration to 13.0

Please @joao-p-marques and @CarlosRoca13 can you review it?

Locked by:
- [x] https://github.com/OCA/reporting-engine/pull/504

@Tecnativa TT26758

### CODE REVIEW

**Comment by feketemihai** [comment]
@victoralmau I think here you just have to check the diff from #150 , probably small migration script, but in the rest should work without the extra code in account.move.

**Comment by victoralmau** [comment]
Thanks @feketemihai, i have changed it now about your initial ideas.
I have add you as co-author in commit migration.

**Review by joao-p-marques — APPROVED** [review]
Code and functional review :+1:

**Review by CarlosRoca13 — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Please include the migration scripts for recovering the comment fields from `account.invoice` to `account.move`. Example script to inspire on:

https://github.com/OCA/l10n-spain/blob/13.0/l10n_es_aeat_sii_oca/migrations/13.0.1.0.0/post-migration.py

**Comment by victoralmau** [comment]
Migration scripts added.
In other similar addon `sale_comment_template` will be necessary too.

**Comment by pedrobaeza** [comment]
@victoralmau please follow the example I have given to you, not other code, and avoiding ORM processing. Also take into account that this is for OpenUpgrade, not for regular updates (that are more debatable about the need of the library), so it's totally OK (and advised) to use `openupgradelib`.

File: account_invoice_comment_template/models/account_move.py
Comment (feketemihai) [inline account_invoice_comment_template/models/account_move.py]: Why do you still keep the code in here, isn't it easier to treat this in the migration script to include also partners. Like this you can remove the code, tests, views, translations...

Comment (victoralmau) [reply account_invoice_comment_template/models/account_move.py]: What do you think about it @pedrobaeza ?

Comment (pedrobaeza) [reply account_invoice_comment_template/models/account_move.py]: I think the approach is incorrect: we still need the specific custom texts on each document, as they can modify them.

**Comment by victoralmau** [comment]
It's ready to merge right?
Ping @pedrobaeza

File: account_invoice_comment_template/__manifest__.py
Comment (ernestotejeda) [inline account_invoice_comment_template/__manifest__.py]:
```suggestion
    "author": "Camptocamp, Tecnativa, Odoo Community Association (OCA)",
```

File: account_invoice_comment_template/models/account_move.py
Comment (ernestotejeda) [inline account_invoice_comment_template/models/account_move.py]: Is it needed?

File: account_invoice_comment_template/models/account_move.py
Comment (ernestotejeda) [inline account_invoice_comment_template/models/account_move.py]: Shouldn't you take the invoice company into account and order by priority?, get_comment_template method of  comment.template model takes into account the company of the model and as i can see, base.comment.template model doesn't have the _order attribute defined.

File: account_invoice_comment_template/models/account_move.py
Comment (ernestotejeda) [inline account_invoice_comment_template/models/account_move.py]: Is that function call necessary? I think the assignment of the previous line should already trigger the self_set_note1 method (since it is a onchange method tha depends on comment_template1_id), what do you think?

File: account_invoice_comment_template/models/account_move.py
Comment (ernestotejeda) [inline account_invoice_comment_template/models/account_move.py]: The same here.

File: account_invoice_comment_template/tests/test_account_invoice_report.py
Comment (ernestotejeda) [inline account_invoice_comment_template/tests/test_account_invoice_report.py]: I think it's not responsibility of this module to test that a report prints its values correctly. Also, if this module is tested together with another module that hides those texts (via qweb) that are being tested here, then the test may fail.

File: account_invoice_comment_template/models/account_move.py
Comment (victoralmau) [reply account_invoice_comment_template/models/account_move.py]: Thanks, i have changed to add `company_id` in domain and order by priority.
It's not possible to use `get_comment_template` function because need `base.comment.template` record and these function not return it. Maybe i have to create a little PR about it.

Comment (victoralmau) [reply account_invoice_comment_template/models/account_move.py]: With these change https://github.com/OCA/reporting-engine/pull/497/ it's possible to change it something like these:

```
_name = "account.move"
_inherit = ["account.move", "comment.template"]

self.get_comment_template_records(
    position="before_lines", company_id=self.company_id.id, partner_id=self.partner_id.id
)[0]
```

**Review by ernestotejeda — APPROVED** [review]

File: account_invoice_comment_template/models/account_move.py
Comment (ernestotejeda) [reply account_invoice_comment_template/models/account_move.py]: I think that would be good. I have already approved that PR.

File: account_invoice_comment_template/__manifest__.py
Comment (pedrobaeza) [inline account_invoice_comment_template/__manifest__.py]: This dependency is not correct. We must do tests that don't force us to install a specific CoA.

File: account_invoice_comment_template/migrations/13.0.1.0.0/post-migration.py
Comment (pedrobaeza) [inline account_invoice_comment_template/migrations/13.0.1.0.0/post-migration.py]: Don't do this subquery, but join the table in the upper query.

File: account_invoice_comment_template/migrations/13.0.1.0.0/post-migration.py
Comment (pedrobaeza) [inline account_invoice_comment_template/migrations/13.0.1.0.0/post-migration.py]: Don't do subquery.

File: account_invoice_comment_template/migrations/13.0.1.0.0/post-migration.py
Comment (pedrobaeza) [inline account_invoice_comment_template/migrations/13.0.1.0.0/post-migration.py]: Don't do subquery.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by pedrobaeza** [comment]
This should be adapted now to the new many2many reality: the fields in the invoice should be also m2m, and filter them by company.

File: account_invoice_comment_template/models/account_move.py
Comment (joao-p-marques) [inline account_invoice_comment_template/models/account_move.py]: I think you will need to give explicit names to this relations' columns, otherwise the ORM will just generate them from the model names, and as there are 2 relations between the same modules, it will cause the conflict in https://travis-ci.com/github/OCA/account-invoice-reporting/jobs/500445740#L1133

Also, as you will already need migration scripts, why not giving a more distinct name to these, like `comment_template_top_ids` and `comment_template_bottom_ids`?

**Review by joao-p-marques — CHANGES_REQUESTED** [review]

**Comment by pedrobaeza** [comment]
This PR must be changed to the latest changes on the base. Please check the base first and dismiss this one until the other is merged and this one adapted.

**Comment by victoralmau** [comment]
According to 14.0, is correct rename addon to `account_comment_template`, right?

**Comment by pedrobaeza** [comment]
Well, I don't see incorrect the current name, but if the renaming has been done, coming back is going to be awful.

**Comment by victoralmau** [comment]
It's ready to review now.

File: account_invoice_comment_template/migrations/13.0.1.0.0/post-migration.py
Comment (pedrobaeza) [inline account_invoice_comment_template/migrations/13.0.1.0.0/post-migration.py]: Never drop columns. Remember: preservation principle. The only exception is if that data compromises integrity, which violates other principle. Even more, you will get an error, as they are m2m fields, not SQL columns. You can lift constraints in such tables for avoiding performance lost though.

File: account_invoice_comment_template/migrations/13.0.1.0.0/pre-migration.py
Comment (pedrobaeza) [inline account_invoice_comment_template/migrations/13.0.1.0.0/pre-migration.py]: This renaming is not preventing the m2m tables linking, so it doesn't serve. Even more, you don't want to rename fields, only columns (which is not the case), when renaming for preserving content in other place. The question here is: the m2m table has the same name? If that's the case, then you should rename the m2m table.

File: account_invoice_comment_template/models/account_move.py
Comment (pedrobaeza) [inline account_invoice_comment_template/models/account_move.py]: Why not removing it? We can discuss in v14 for its removal, but I would remove it here for now.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: account_invoice_comment_template/migrations/13.0.1.0.0/post-migration.py
Comment (victoralmau) [reply account_invoice_comment_template/migrations/13.0.1.0.0/post-migration.py]: Sorry for my mistake, the name of the fields was wrong (it is not an m2m, it is 2 m2o fields that convert to m2m).

Comment (pedrobaeza) [reply account_invoice_comment_template/migrations/13.0.1.0.0/post-migration.py]: You are still dropping columns and this is not the mentioned exception.

File: account_invoice_comment_template/migrations/13.0.1.0.0/post-migration.py
Comment (pedrobaeza) [inline account_invoice_comment_template/migrations/13.0.1.0.0/post-migration.py:30]: Add `ON CONFLICT DO NOTHING`, as there can be the case where you have the same template on top and on bottom.

File: account_invoice_comment_template/migrations/13.0.1.0.0/pre-migration.py
Comment (pedrobaeza) [inline account_invoice_comment_template/migrations/13.0.1.0.0/pre-migration.py]: Why are you renaming columns if the previous name is not overlapped by anything? The only reason for renaming columns is for avoiding to conflict with the same name, and as it's very unlikely that other module uses this field name, it's better to save on scripts doing the renaming + avoid to use the renamed one on further scripts. Anyways, if needed, the usual technique is to rename to `None`, for letting OpenUpgrade to put the name with `openupgradelib.get_legacy_name`.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**
